Coccinelle updates of unsigned variable assignments

pfl opened the pull request:
Here is a Coccinelle script that adds a 'U' to unsigned variable assignments according to MISRA-C rule 7.2. and issue #9895. The script can report where a 'U' is missing and also propose a patch. The report part can be run by each developer and could also be added to CI. Looking at the analysis tool, 777 instances have now shrunk down to 47, which need to be investigated still.

Patches for each directory except ./ext are attached.

pfl committed:
scripts/coccinelle: Add Coccinelle script for unsigned values

Add a Coccinelle script that adds an 'U' to values assigned to
unsigned variables, according ot MISRA-C rule 7.2.

Add a 'report' mode to the script that can be used by developer/CI
and a 'patch' mode that should do the heavy lifting.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

ceolin reviewed (approved):
Patch is awesome ! Just a comment, don't you think is better having only one commit for code changes in this case ? I mean since everything was changed using the same script, I rather prefer having one commit with the changes so if we find a problem is easier to revert.

nashif commented:
> Patch is awesome ! Just a comment, don't you think is better having only one commit for code changes in this case ? I mean since everything was changed using the same script, I rather prefer having one commit with the changes so if we find a problem is easier to revert.

how else would we reach 30k commit by end of the year? :-)

ceolin commented:
> > Patch is awesome ! Just a comment, don't you think is better having only one commit for code changes in this case ? I mean since everything was changed using the same script, I rather prefer having one commit with the changes so if we find a problem is easier to revert.
> 
> how else would we reach 30k commit by end of the year? :-)

lol, but honestly split into several commits changes caused by scripts does not scale, when we do naming changes it can easily go over 100 commits, it is insane. I think it is better one commit introducing the script and one commit with the changes, then it's easy to revert if needed.

pfl commented:
> lol, but honestly split into several commits changes caused by scripts does not scale, when we do naming changes it can easily go over 100 commits, it is insane. I think it is better one commit introducing the script and one commit with the changes, then it's easy to revert if needed.

Well, if it is split into multiple parts, it is human readable as well. People may be able to scroll to the end this way. And should someone object, only a part of the changes are shot down :-)

jarz-nordic commented:
What about commonly used types like u8_t u_16t? I did not see them in the script.

Great job!

pfl commented:
> What about commonly used types like u8_t u_16t? I did not see them in the script.

That's what the remaining ones are lurking behind! Seems I didn't remember which unsigned definition was in use in Zephyr. I'll update.

himanshujha199640 commented on scripts/coccinelle/unsigned_assignment.cocci:
```diff
@@ -0,0 +1,40 @@
+// Find assignments to unsigned variables and add an 'U' to the value
```
https://docs.zephyrproject.org/latest/application/coccinelle.html#proposing-new-semantic-patches

Use `///` here.

And after doing all changes run(assuming you've installed coccinelle):
`$./scripts/coccicheck --mode=patch --verbose=1 --cocci=scripts/coccinelle/unsigned_assignment.cocci`

and  you will see the message:
`Find assignments to unsigned variables and add an 'U' to the value`

Also add `Confidence` which seems HIGH since no false positives are observed.

himanshujha199640 commented on scripts/coccinelle/unsigned_assignment.cocci:
```diff
@@ -0,0 +1,40 @@
+virtual patch
+virtual report
+
+@r_unsigned@
```
add `depends on !(file in "ext")` to all rules to exclude processing
`ext/` directory from processing.

Ref: https://github.com/zephyrproject-rtos/zephyr/commit/817f4b374fba536db2740fdc8965e35a88ac5e2d

pfl replied:
When I add this, spatch version 1.0.4.deb-3+b4 from Debian testing complains with:
Fatal error: exception Failure("rule name: parse error: \n = File \".../zephyr/scripts/coccinelle/unsigned_assignment.cocci\", line 11, column 25,  charpos = 311\n    around = 'file', whole content = @r_unsigned depends on !(file in \"ext\")@\n")

It seems I need to test if a newer version of coccinelle/spatch works better.

himanshujha199640 replied:
Hmm.
Well, 1.0.4 is too old and we always recommend to use latest version:
```
himanshu@himanshu-Vostro-3559:~/coccinelle$ git diff --shortstat master 1.0.4 
 829 files changed, 18373 insertions(+), 67399 deletions(-)
```
I updated all scripts to use `depends on !(file in "ext")` since `ext/` comes from 
third paty vendor repo and there was less interest for those reports.

And it works fine for me.

himanshujha199640 commented on scripts/coccinelle/unsigned_assignment.cocci:
```diff
@@ -0,0 +1,40 @@
+position p;
+@@
+
+  v = C@p
```
Use:
```
(
v = i 
|
v = C 
)
```
where i is an identifier
and get rid of the isdigit() check as @JuliaLawall suggested!

We use python/Ocaml scripting when we have no options SmPL grammar.

pfl replied:
This modification works ok, but...

himanshujha199640 commented on scripts/coccinelle/unsigned_assignment.cocci:
```diff
@@ -0,0 +1,40 @@
+@@
+
+if C.isdigit() != True:
+    cocci.include_match(False)
```
Now get rid of these two lines above.

pfl replied:
...removing the isdigit() check down here causes the patch mode to add an extra 'U' in ext/ like so:
`diff -u -p a-project/ext/hal/cmsis/Include/mpu_armv8.h b-project/ext/hal/cmsis/Include/mpu_armv8.h
--- a-project/ext/hal/cmsis/Include/mpu_armv8.h
+++ b-project/ext/hal/cmsis/Include/mpu_armv8.h
@@ -270,7 +270,7 @@ __STATIC_INLINE void ARM_MPU_SetRegion_N
 __STATIC_INLINE void orderedCpy(volatile uint32_t* dst, const uint32_t* __RESTRICT src, uint32_t len)
 {
   uint32_t i;
-  for (i = 0U; i < len; ++i) 
+  for (i = 0UU; i < len; ++i) 
   {
     dst[i] = src[i];
   }
`
I have yet to figure out whether this too is caused by version 1.0.4.

himanshujha199640 replied:
No, it isn't caused by 1.0.4!

```
himanshu@himanshu-Vostro-3559:~/zephyr$ cat test.cocci test.c; spatch --sp-file test.cocci test.c
@r@
{unsigned short, unsigned int} v;
constant C;
@@

* v = C

int main (void)
{
	unsigned int foo = 20U;
}
init_defs_builtins: /usr/local/bin/../lib/coccinelle/standard.h
HANDLING: test.c
diff = 
--- test.c
+++ /tmp/cocci-output-5327-3dff8c-test.c
@@ -1,5 +1,4 @@
 
 int main (void)
 {
-	unsigned int foo = 20U;
 }
```

Instead `constant` matches `0U` in this case and adds an extra `U`.
So, isdigit() hack is good to go unless we have any methods to
check for an existing `U`.

himanshujha199640 commented on scripts/coccinelle/unsigned_assignment.cocci:
```diff
@@ -0,0 +1,40 @@
+@script: python depends on report@
+p << r_unsigned.p;
+@@
+print "Unsigned 'U' suffix missing on line %s of %s" % (p[0].line, p[0].file)
```
See how report is handled at `scripts/coccinelle/mini_lock.cocci`
and use the same format.

pfl replied:
Ok.

himanshujha199640 reviewed (changes requested):
Thanks for using Coccinelle.
Few minor bits from me.

Cc @JuliaLawall

pfl commented:
Let's go with the changes that work on at least Debian's version of Coccinelle 1.0.4 and fight/fix the script later to be nicer. The updated script takes care of the uXX_t unsigned definitions.

himanshujha199640 commented on scripts/coccinelle/unsigned_assignment.cocci:
```diff
@@ -0,0 +1,44 @@
+p << r_unsigned.p;
+@@
+
+coccilib.report.print_report(p[0], "Unsigned 'U' suffix missing")
```
I would prefer and follow the tradition:
```
msg="WARNING: Unsigned 'U' suffix missing"
coccilib.report.print_report(p[0], msg)
```

Also, what do you think about naming the file as `unsigned_suffix.cocci` or
something like that ?

pfl replied:
Yes, warning printout fixed and scrip name has been updated.

aescolar reviewed (commented):
No complaints for the native_posix files. (did not review the rest)

pfl committed:
arch: Add 'U' to unsigned variable assignments

Add 'U' to a value when assigning it to an unsigned variable.
MISRA-C rule 7.2

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
boards: Add 'U' to unsigned variable assignments

Add 'U' to a value when assigning it to an unsigned variable.
MISRA-C rule 7.2

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
drivers: Add 'U' to unsigned variable assignments

Add 'U' to a value when assigning it to an unsigned variable.
MISRA-C rule 7.2

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
kernel: Add 'U' to unsigned variable assignments

Add 'U' to a value when assigning it to an unsigned variable.
MISRA-C rule 7.2

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
lib: Add 'U' to unsigned variable assignments

Add 'U' to a value when assigning it to an unsigned variable.
MISRA-C rule 7.2

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
misc: Add 'U' to unsigned variable assignments

Add 'U' to a value when assigning it to an unsigned variable.
MISRA-C rule 7.2

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
soc: Add 'U' to unsigned variable assignments

Add 'U' to a value when assigning it to an unsigned variable.
MISRA-C rule 7.2

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
tests: Add 'U' to unsigned variable assignments

Add 'U' to a value when assigning it to an unsigned variable.
MISRA-C rule 7.2

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
samples: Add 'U' to unsigned variable assignments

Add 'U' to a value when assigning it to an unsigned variable.
MISRA-C rule 7.2

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
misra-c: Add 'U' to unsigned variable assignments in subsys/

Add 'U' to a value when assigning it to an unsigned variable.
MISRA-C rule 7.2

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl commented:
Rebased.

himanshujha199640 reviewed (changes requested):
One minor comment which I observed while
testing.

Cc @JuliaLawall

pfl commented on scripts/coccinelle/unsigned_suffix.cocci:
```diff
@@ -0,0 +1,45 @@
+
+coccinelle.z = C + "U"
+
+@r_subst@
```
Done.

himanshujha199640 reviewed (approved):
LGTM!

The shippable failure seems related to: https://github.com/zephyrproject-rtos/zephyr/issues/11754
There is also one merge conflict so you need to rebase.

Note that this script lacks `depends on !(file in "ext")` so running this would
produce warnings for `ext/` too which I mentioned in a previous review
request to this PR.

Thanks!

pfl commented:
> Note that this script lacks `depends on !(file in "ext")` so running this would
> produce warnings for `ext/` too which I mentioned in a previous review
> request to this PR.

This had some issues with coccinelle 1.0.4, let's try to get these changes in and then fix the script to saner behavior.

JuliaLawall commented:
On Fri, 30 Nov 2018, Patrik Flykt wrote:

>
>       Note that this script lacks depends on !(file in "ext") so
>       running this would
>       produce warnings for ext/ too which I mentioned in a previous
>       review
>       request to this PR.
>
> This had some issues with coccinelle 1.0.4, let's try to get these changes
> in and then fix the script to saner behavior.

I don't think file in is supported in 1.0.4.  I will see if the Ubuntu PPA
can be updated to 1.0.7, if that would help.

https://launchpad.net/~npalix/+archive/ubuntu/coccinelle

julia

himanshujha199640 commented:
On Fri, Nov 30, 2018 at 07:33:26PM +0000, JuliaLawall wrote:
> 
> 
> On Fri, 30 Nov 2018, Patrik Flykt wrote:
> 
> >
> >       Note that this script lacks depends on !(file in "ext") so
> >       running this would
> >       produce warnings for ext/ too which I mentioned in a previous
> >       review
> >       request to this PR.
> >
> > This had some issues with coccinelle 1.0.4, let's try to get these changes
> > in and then fix the script to saner behavior.
> 
> I don't think file in is supported in 1.0.4.  I will see if the Ubuntu PPA
> can be updated to 1.0.7, if that would help.
> 
> https://launchpad.net/~npalix/+archive/ubuntu/coccinelle

Yes, excellent idea! Please try if possible.

I often find developers reluctant to install from source and
most of them prefer hassle free distribution packages.

At least after that I can convince more people to try it, though
majority use fedora for development.

Also, IIRC few of linux wireless folks do use the aforementioned
PPA.

-- 
Himanshu Jha
Undergraduate Student
Department of Electronics & Communication
Guru Tegh Bahadur Institute of Technology

JuliaLawall commented:
On Fri, 30 Nov 2018, Himanshu Jha wrote:

> On Fri, Nov 30, 2018 at 07:33:26PM +0000, JuliaLawall wrote:
> >
> >
> > On Fri, 30 Nov 2018, Patrik Flykt wrote:
> >
> > >
> > > Note that this script lacks depends on !(file in "ext") so
> > > running this would
> > > produce warnings for ext/ too which I mentioned in a previous
> > > review
> > > request to this PR.
> > >
> > > This had some issues with coccinelle 1.0.4, let's try to get these changes
> > > in and then fix the script to saner behavior.
> >
> > I don't think file in is supported in 1.0.4. I will see if the Ubuntu PPA
> > can be updated to 1.0.7, if that would help.
> >
> > https://launchpad.net/~npalix/+archive/ubuntu/coccinelle
>
> Yes, excellent idea! Please try if possible.

I asked, but no response so far.  Perhaps on Monday.

julia

>
> I often find developers reluctant to install from source and
> most of them prefer hassle free distribution packages.
>
> At least after that I can convince more people to try it, though
> majority use fedora for development.
>
> Also, IIRC few of linux wireless folks do use the aforementioned
> PPA.
>
> --
> Himanshu Jha
> Undergraduate Student
> Department of Electronics & Communication
> Guru Tegh Bahadur Institute of Technology
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.[AAesmg0dK_5kvr9Ym8aMCi7yRTsA2i7-ks5u0Y5FgaJpZM4YmvOX.gif]
>
>
>

pfl commented:
Another day, another rebase. If we could quickly push this it'd be great, else I start thinking of splitting this into several parts - as its quite large (but trivial), keeping everything without conflicts is an ever evolving game. @andrewboie @andyross ?
